scalyr_agent/date_parsing_utils.py

```diff
@@ -0,0 +1,465 @@
+
+    return nano_seconds + nanos
+
+
+def _rfc3339_to_nanoseconds_since_epoch_regex(string):
+    """
+    rfc3339_to_nanoseconds_since_epoch variation which utilizes regex approach
+    """
+    # split the string in to main time and fractional component
+    parts = string.split(".")
+
+    # it's possible that the time does not have a fractional component
+    # e.g 2015-08-03T09:12:43Z, in this case 'parts' will only have a
+    # single element that should end in Z.  Strip the Z if it exists
+    # so we can use the same format string for processing the main
+    # date+time regardless of whether the time has a fractional component.
+    if parts[0].endswith("Z"):
+        parts[0] = parts[0][:-1]
+
+    # NOTE: We benchmarked multiple versions, including regex one and using string.split
+    # appears to be a bit faster than regex
+    try:
+        dt = datetime.datetime(
+            *list(map(int, RFC3339_STR_REGEX.match(string).groups()))
+        )
+    except Exception:
+        return None
+
+    nano_seconds = (
+        calendar.timegm((dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second))
+        * 1000000000
+    )
+
+    nanos = 0
+
+    # now add the fractional part
+    if len(parts) > 1:
+        fractions = parts[1]
+        # if the fractional part doesn't end in Z we likely have a
+        # malformed time, so just return the current value
+        if not fractions.endswith("Z"):
+            # we don't handle non UTC timezones yet
+            if any(c in fractions for c in "+-"):
+                return None
+
+            return nano_seconds
+
+        # strip the final 'Z' and use the final number for processing
+        fractions = fractions[:-1]
+        to_nanos = 9 - len(fractions)
+        nanos = int(int(fractions) * 10 ** to_nanos)
+
+    return nano_seconds + nanos
+
+
+def _rfc3339_to_nanoseconds_since_epoch_string_split(string):
+    """
+    rfc3339_to_nanoseconds_since_epoch variation which utilizes string.split approach
+    """
+    # split the string in to main time and fractional component
+    parts = string.split(".")
+
+    # it's possible that the time does not have a fractional component
+    # e.g 2015-08-03T09:12:43Z, in this case 'parts' will only have a
+    # single element that should end in Z.  Strip the Z if it exists
+    # so we can use the same format string for processing the main
+    # date+time regardless of whether the time has a fractional component.
+    if parts[0].endswith("Z"):
+        parts[0] = parts[0][:-1]
+
+    # NOTE: We benchmarked multiple versions, including regex one and using string.split
+    # appears to be a bit faster than regex
+    try:
+        # NOTE: I intentionally access values directly in the list and don't assign them
+        # to intermediate variables since it's faster
+        result = parts[0].split("T")
+        result1 = result[0].split("-")
+        result2 = result[1].split(".")[0].split(":")
```

Comment:
I believe you are correct - that split is not needed.

Reply:
Here we go - d2cfd726350d639ea44c7b442e6d039b0fe65dae.

Also covers the other comment.

Reply:
And with that unnecessary split removed, split string version of ``_to_datetime`` function is now as fast as the udatetime version.

In fact, locally it sometimes finishes faster than udatetime version (although the difference is indeed very small).

Keep in mind this is only true for ``_to_datetime`` version of this function. For ``_to_nanoseconds_seconds_since_epoch`` udatetime is still faster.

The reason for that is that in this second version, we still need to call ``calendar.timegm`` which adds additional overhead which is not needed when using udatetime version.

This also shows that native Python implementation can sometimes also be faster than the C one :)

But this makes sense in this case since ``udatetime`` supports other more complex scenarios (timezones) which we don't need to support.

Reply:
> And with that unnecessary split removed, split string version of _to_datetime function is now as fast as the udatetime version.

Nice! 👍  How close is the difference?  If it's not going have a significant impact on the hot path, I'd almost be inclined *not* to use udatetime, so as to keep the code simple and not to introduce another dependency.

Reply:
In ``rfc3339_to_nanoseconds_since_epoch`` method the difference is 30% or so, so it's likely worth it.

In other method case, the difference is much smaller (just a couple of % faster), but I think it's worth to leave it in place because to the "to_nanosecond" method.

It will still work out of the box, if the library is not available we will just fall back to the native approach.

In short, we will follow the same approach as we do with ujson - we won't bundle it, but we will start a documentation page which will instruct advanced users how to squeeze out some additional performance by changing various config option and installing libraries such as ujson and u datetime.